wg-meetings/2015-11-19.md

```diff
@@ -0,0 +1,168 @@
+
+
+Public and Private APIs
+--------------------------------------------------------------------------------
+
+[YouTube 3:56](https://youtu.be/3rszjGYpbyM?t=236)
+
+Oguz:
+
+General priorities:  
+
+1. private - reduce amount of work between native bits using shims
+2. public - native add-ons; nan as an example
+
+Public API could have it's own types
+
+Patrick: we need some concrete issues
+
+Michael: What APIs are we talking about - the API we provide to modules?
+
+Oguz: yes
+
+Michael: need a full definition
+
+Patrick: different forms of "definition"; C code, documentation, API
+modelling languages, ...
+
+Michael: Oguz's current one is in C?
+
+Oguz: current JxCore API isn't appropriate for public API
+
+Rob: current API is V8; we should audit common modules to figure out what APIs
+we need
+
+Trevor: still limitations; special VM APIs like GC notification
+
+Patrick: maybe 80/20 is good enough for now
+
+Trevor: how would this be "blessed", and if it has a performance impact, will
+never be blessed
+
+Rob: should AIM to be macros and templates aimed at V8, so no performance impact;
+other impls might need to add C code
+
+Patrick: let's create an issue for this
+
+Oguz: 60% of this will be easy
+
+Trevor: this approach is the most complicated way to get anything done; need to
+also get the all the existing modules converted
+
+Rob: no existing modules are going to work with a new API anyway
+
+Trevor: it's getting complicated; now we need to support multiple APIs
+
+Oguz: SQLite is probably a good example of types of APIs external modules need
+
+Patrick: we need an issue
+
+Trevor: does other native headers (Chakra) exist in V8?  Do we ship multiple
+versions of node?
+
+Rob: no other native headers; runtimes act like drivers
+
+Patrick: we need an issue
+
+Rob: I volunteer *[pjm: not his exact words :-) ]*
+
+Trevor: more actionable if consolidate we node header into a single header;
+one official header; also create a user-facing shim for V8, not specifically
+targeted for abstraction, just to get started
+
+Patrick: what does the user-facing library look like?
+
+Trevor/Michael/Rob: like nan, more API and ABI stable than nan, but can never be
+100% API and especially ABI stable, and can be included in node core.  For
+native module authors.  Eventually use that API to abstract over multiple JS
+implementations.
+
+Patrick: I think we're talking about two main issues: abstracting JS engines and
+API for module authors.  Same problem space, attacking from different ends.
+Let's get some issues created.
+
+
+FFI
+--------------------------------------------------------------------------------
+
+[YouTube 38:44](https://youtu.be/3rszjGYpbyM?t=2324)
+
+Alexis - FFI is invoking C functions from JS, most native modules could use
+this to do their work.  Some issues have come up regarding performance issues,
+but there are some approaches which minimize performance, such as
+Chakra using templates and having a compiler provide some glue.
```

Comment:
This is not actually used by Chakra, though it was suggested by a member of the Chakra team. We can just say `as suggested at https://github.com/nodejs/nan/issues/349#issuecomment-110569177`.